utils/makefile.python

```diff
@@ -99,7 +99,9 @@ quiet_cmd_pyinstall = INSTALL   $2
 quiet_cmd_pyenvinstall = PYENV     install $2
       cmd_pyenvinstall = \
 	if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then \
-		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) ;\
+		rm -f $(PY_ENV)/requirements.sha256; \
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install $$(grep pyyaml ./requirements.txt); \
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) &&\
```

Comment:
Thought: if I ignore the git workflow caching issue and have a complete new look at this two lines, I cant understand why line 103 is needed .. or have I lost the overview?

Reply:
https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/Makefile#L185

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/Makefile#L107-L108

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/utils/build_env.py#L19

brand is variable initialized in `searx/__init__.py` :

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/searx/__init__.py#L27

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/searx/settings_loader.py#L8

Reply:
Sorry I am an idiot .. I totally forgot the main problem

Reply:
Hm .. after thinking again .. pyyaml  should already be installed when the virtualenv is build

https://github.com/dalf/searx/blob/fe3085d68e965732ca7291a14225b34a7e046e92/utils/makefile.python#L122-L129

and this will be installed before pyenvinstall is called:

https://github.com/dalf/searx/blob/fe3085d68e965732ca7291a14225b34a7e046e92/utils/makefile.python#L193

In https://github.com/searx/searx/pull/2517#discussion_r567049019 I described it:

> pyenv: builds the environment for development / typically `requirements.txt`
> penvinstall: installs the python package in this environment (`pip install -e`) / typically `setup.py` 

The other point you asked:

> each invocation of pyenv will reinstall the virtualenv

Our mistake was to build `sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256 ;\` in ` cmd_pyenvinstall` .. this optimizing should take place in `cmd_virtualenv`.

Or I am wrong? .. I will give it a try ...

Reply:
> Hm .. after thinking again .. pyyaml should already be installed when the virtualenv is build

cmd_virtualenv seems to fail sometimes on github action:
* in this case, the ./local contains an initialized virtualenv, but without pyyaml
* cmd_virtualenv won't detect that pyyaml is missing (or other dependencies): the virtualenv exists, `...python` is an executable.
* then cmd_pyenvinstall fails because pyyaml is missing.
* the cache is updated with a broken virtualenv.
* all new builds on github action fail.

An alternative implementation (without `...grep pyyaml ./requirements.txt)...` ) :
* after `$(PY_ENV_BIN)/python -m pip install $(PIP_VERBOSE) -r requirements.txt;`, a file is created (or just a copy of requirements.txt)
* `if [ -d "./$(PY_ENV)" -a -x "./$(PY_ENV_BIN)/python" ]; then` also check if this file exists.
* basically it duplicates `if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then` in cmd_virtualenv.

Reply:
I work on a solution .. give me some time this day and I will show you my solution:

> An alternative implementation 
> ...
> ..

All this goes in the right direction to hardening, but instead hardening the installation, we have to hardening the venv build .. I will consider it in my solution / thanks!

Reply:
@dalf I implemented a alternative PR #2524 ... its all described there ..